plugins/outputs/application_insights/application_insights.go

```diff
@@ -0,0 +1,410 @@
+	return "Send telegraf metrics to Azure Application Insights"
+}
+
+func (a *ApplicationInsights) Connect() error {
+	if a.Timeout.Duration == 0 {
+		a.Timeout = internal.Duration{Duration: time.Second * 5}
+	}
+
+	if a.Transmitter == nil {
+		a.useDefaultTransmitter()
+	}
+
+	if a.Transmitter != nil && a.DiagMsgSubscriber != nil {
+		a.diagMsgListener = a.DiagMsgSubscriber.Subscribe(func(msg string) error {
+			logOutputMsg("%s", msg)
+			return nil
+		})
+	}
+
+	return nil
+}
+
+func (a *ApplicationInsights) Write(metrics []telegraf.Metric) error {
+	if len(metrics) == 0 || a.Transmitter == nil || a.InstrumentationKey == "" {
+		return nil
+	}
+
+	for _, metric := range metrics {
+		if canBeTreatedAsAggregate(metric) {
+			if telemetry := createAggregateMetricTelemetry(metric); telemetry != nil {
+				a.Transmitter.Track(telemetry)
+				continue
+			}
+		}
+
+		if telemetry := createSimpleMetricTelemetry(metric); telemetry != nil {
+			a.Transmitter.Track(telemetry)
+		}
+	}
+
+	return nil
+}
+
+func (a *ApplicationInsights) Close() error {
+	if a.Transmitter == nil {
+		return nil
+	}
+
+	select {
+	case <-a.Transmitter.Close():
+		logOutputMsg("Closed")
+	case <-time.After(a.Timeout.Duration):
+		logOutputMsg("Close operation timed out after %v", a.Timeout.Duration)
+	}
+
+	if a.diagMsgListener != nil {
+		// We want to listen to diagnostic messages during closing
+		// That is why we stop listening only after Close() ends (or a timeout occurs)
+		a.diagMsgListener.Remove()
+	}
+
+	return nil
+}
+
+func init() {
+	outputs.Add("application_insights", func() telegraf.Output {
+		return &ApplicationInsights{
+			DiagMsgSubscriber: appinsightsDiagnosticsMessageSubscriber{},
+		}
+	})
+}
+
+func (a *ApplicationInsights) useDefaultTransmitter() {
+	// Env variable, if present, overrides what is in the configuration
+	iKey := os.Getenv("APPINSIGHTS_INSTRUMENTATIONKEY")
+	if iKey != "" {
+		a.InstrumentationKey = iKey
+	}
+
+	if a.InstrumentationKey != "" {
+		a.Transmitter = NewAppinsightsTransmitter(a.InstrumentationKey)
+	}
+}
+
+func canBeTreatedAsAggregate(metric telegraf.Metric) bool {
+	// If we can identify the Sum and Count properties, then we are good
+	hasSum := metric.HasField("value") || metric.HasField("sum")
+	hasCount := metric.HasField("count") || metric.HasField("samples")
+	return hasSum && hasCount
+}
+
+func createSimpleMetricTelemetry(metric telegraf.Metric) *appinsights.MetricTelemetry {
+	telemetry := appinsights.NewMetricTelemetry(metric.Name(), 0.0)
+	telemetry.Properties = metric.Tags()
+	telemetry.Timestamp = metric.Time()
+
+	err := setFloat64TelemetryProperty(&(telemetry.Value), []string{"value"}, metric, nil)
+	if err == nil {
+		addRemainingMeasurements(telemetry, metric, []string{"value"})
+	} else {
+		// Either the metric does not have a "value" field, or it was not convertable to float64
+		// Heuristics: if the metric has just one field that is convertible to float64, take it as value
+		// Otherwise give up
```

Comment:
I think the "value" field is just a special case. And the general case is metric with multiple dimensions (fields), and we shouldn't give up in this case.
For example, I can have a metric like
{
    metricName: "cpu",
    fields: [
        "Percent_Idle_Time": 0.3,
        "Percent_User_Time": 0.5
    ]
}
and we can generate multiple AI telemetries from this metric, with metric names cpu_Percent_Idle_Time, cpu_Percent_User_Time

Reply:
@yantang-msft That's correct.  A measurement (in telegraf/influx) could have multiple fields.  

The logic here would would only pick up a measurement that happens to have a `value` field and is convertible to a `float64`... dropping all other fields which is unlikely what is intended.

Reply:
Agreed, will submit new PR to fix this